gateway/ln-gateway/src/lib.rs

```diff
@@ -1611,6 +1615,73 @@ impl Gateway {
         Ok(())
     }
 
+    /// Queries the client log for transaction events and returns to the user.
+    pub async fn handle_transactions_msg(
+        &self,
+        TransactionsPayload {
+            position,
+            limit,
+            federation_id,
+            event_kinds,
+        }: TransactionsPayload,
+    ) -> TransactionsResponse {
+        let federation_manager = self.federation_manager.read().await;
+        let configs = {
+            let mut dbtx = self.gateway_db.begin_transaction_nc().await;
+            dbtx.load_federation_configs().await
+        };
+
+        let clients = configs
+            .into_iter()
+            .filter_map(|(fed_id, _)| match federation_id {
+                Some(id) if id == fed_id => Some((
+                    id,
+                    federation_manager
+                        .client(&fed_id)
+                        .expect("Invalid federation")
+                        .value(),
+                )),
+                Some(_) => None,
+                None => Some((
+                    fed_id,
+                    federation_manager
+                        .client(&fed_id)
+                        .expect("Invalid federation")
+                        .value(),
+                )),
+            })
+            .collect::<Vec<_>>();
+
+        let event_kinds = if event_kinds.is_empty() {
+            vec![
+                OutgoingPaymentSuccess::KIND,
+                IncomingPaymentSuccess::KIND,
+                CompleteLightningPaymentSuccess::KIND,
+            ]
+        } else {
+            event_kinds
+        };
+
+        let mut transactions = Vec::new();
+
+        for (fed_id, client) in clients {
+            let all_events = client.get_event_log(position, limit).await;
+            let filtered_transactions = all_events
+                .into_iter()
+                .filter_map(|e| {
+                    if event_kinds.contains(&e.1) {
+                        Some((e.0, e.1, e.2, e.3, fed_id, e.4))
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+            transactions.extend(filtered_transactions);
+        }
```

Comment:
This could be written as a stream with flatten and filter_map.

Reply:
Refactored this a bit, let me know if this is what you had in mind. I had trouble pushing the filter inside of the stream but otherwise I think it is an improvement over how it was written before